tracing-subscriber/src/registry/mod.rs

```diff
@@ -291,43 +471,87 @@ where
     }
 }
 
-impl<'a, R> Iterator for Parents<'a, R>
-where
-    R: LookupSpan<'a>,
-{
-    type Item = SpanRef<'a, R>;
-    fn next(&mut self) -> Option<Self::Item> {
-        let id = self.next.take()?;
-        let span = self.registry.span(&id)?;
-        self.next = span.parent().map(|parent| parent.id());
-        Some(span)
-    }
-}
-
-// === impl FromRoot ===
-
-impl<'span, R> Iterator for FromRoot<'span, R>
-where
-    R: LookupSpan<'span>,
-{
-    type Item = SpanRef<'span, R>;
+#[cfg(test)]
+mod tests {
+    use crate::{
+        layer::{Context, Layer},
+        prelude::*,
+        registry::LookupSpan,
+    };
+    use std::sync::{Arc, Mutex};
+    use tracing::{span, Subscriber};
 
-    #[inline]
-    fn next(&mut self) -> Option<Self::Item> {
-        self.inner.next()
-    }
-
-    #[inline]
-    fn size_hint(&self) -> (usize, Option<usize>) {
-        self.inner.size_hint()
+    #[test]
+    fn spanref_scope_iteration_order() {
+        let last_entered_scope = Arc::new(Mutex::new(Vec::new()));
+        #[derive(Default)]
+        struct PrintingLayer {
+            last_entered_scope: Arc<Mutex<Vec<&'static str>>>,
```

Comment:
this works and is fine, but an alternative way of writing these tests could just be to have the layer itself make the assertions. you could construct it with a `VecDeque<Vec<&'static str>>`, with the expected spans at each `on_enter` call in order, and every time `on_enter` is called, you could pop the front of the deque and assert that it matches what you get from `collect`ing the iterator. this _might_ be a bit simpler than sharing state between the layer and the rest of the test, but either approach is fine.

Reply:
It would mean losing the `Arc`, which would be nice. On the other hand, I'm not a huge fan of the way that that decouples the action from the expected outcome.

Reply:
yeah, either way works fine --- it is nice to have the assertions made in the body of the test. i'm happy either way!